jnum_chk.8
.\" section 8 man page for jnum_chk
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2022.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec misfeatures and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH jnum_chk 8 "22 October 2022" "jnum_chk" "IOCCC tools"
.SH NAME
jnum_chk \- tool to check JSON number string conversions
.SH SYNOPSIS
\fBjnum_chk\fP [\-h] [\-v level] [\-J level] [\-V] [\-q] [\-S]
.SH DESCRIPTION
\fBjnum_chk\fP verifies that the JSON number string conversions work correctly, using the data in \fIjnum_test.c\fP which is automatically generated by the \fBjnum_gen(8)\fP tool.
By default only 8, 16, 32, 64 bit and max size signed and unsigned integer types are checked.
Floating point, double floating point and long double floating point are also checked but only to 1 part in 4.1943e+06 (\fBMATCH_PRECISION\fP).
Thus floating point conversions do not have to match fully in order to succeed when strict mode is disabled (the default).
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-v \fIlevel\fP\fP
Set verbosity level to level (def: 0).
.TP
\fB\-V\fP
Print version and exit.
.TP
\fB\-J \fIlevel\fP\fP
Set JSON verbosity level to level (def: 0).
.TP
\fB\-q\fP
Set quiet mode.
This will be passed to the JSON parser.
.PP
\fB\-S\fP
Set strict mode.
.SH EXIT STATUS
.TP
0
all is OK
.TQ
1
Without \-S given and JSON number conversion test suite failed
.TQ
2
\-S given and JSON number conversion test suite failed
.TQ
3
\-h and help string printed or \-V and version string printed
.TQ
4
command line error
.TQ
>=5
internal error
.SH NOTES
.PP
The JSON parser \fBjparse\fP was co\-developed by Cody Boone Ferguson and Landon Curt Noll (one of the IOCCC Judges) in support for IOCCCMOCK, IOCCC28 and beyond.
.PP
The strict mode will enable all tests and force the floating point checks to match fully.
Specifically, if strict mode is disabled (the default) and the two numbers do not match but the difference is <= \fB1.0/MATCH_PRECISION\fP all is OK.
Otherwise there is an issue.
The strict mode option \fB\-S\fP is for informational purposes only.
If it fails on your system this is okay: the mkiocccentry repo does not need \fB\-S\fP to pass in order to be able to create a valid IOCCC entry compressed tarball.
.SH BUGS
If you have a problem with the tool (not JSON itself! :\-) ) you can report it at the GitHub issues page.
It can be found at
.br
\fI\<https://github.com/ioccc\-src/mkiocccentry/issues\>\fP.
.SH SEE ALSO
\fBjparse(1)\fP, \fBjnum_gen(8)\fP.


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.

The `chk_foo()` functions in `chk_validate.c` and the `test_foo()` functions in
`entry_util.c` are code complete, although they remain untested and unused.  The
`chkentry` tool, however, is not yet code complete. Later releases of tested
JSON semantic code will no doubt modify these functions.

Improved a number of the ways that JSON field values are checked.  In a number
of cases, code form `mkiocccentry.c` was moved into `test_foo()` functions so
that they could be used by other tools such as the JSON semantic test code.

Added code to generate JSON semantic tables from JSON reference files for
`.info.json` and `.author.json`.  The `jsemcgen.sh` tool manages this by way of
the `jsemtblgen` code generator and header, patch and trailer files (see
`chk.auth.*` and `chk.info.*` files).

Avoided the appearance of attacking any particular individual.  It was not our
intention to disrespect anyone, even though we disagree with some of the
technical decisions.  Where we have fundamental technical disagreements, we
attempted to express those technical disagreements with humour in hopefully a
more fun way.  As also now apologize for how `bison` and `flex` generated code
may look, instead of simply calling it ugly.  As such, we hopefully improved
some of the humour in our code while trying to be nice and friendly to others.

For example now the adjusted dbg levels in JSON parser are:

```
*     At -J 3, only the top level return type and top level tree are printed.
*     At -J 5, intermediate tree return types and tree are printed too.
*     At -J 7, also print grammar progress.
*     At -J 9, also print sorry_text and sorry_lang grammar values.
```

Removed a number of files and added a number of files under the `test_JSON/`
tree.  When the JSON semantic code is being tested in a future release, we
expect more such `test_JSON/` tree changes.

Improved / add a number of man pages.  Updated `README.md`.

Improved and expanded `txzchk`.

Added more test code.  We attempt to detect feathers in tarballs.  :-)

We will neither confirm nor deny the presence of an "Easter egg".
To to so would be "foolish".  :-)

Improved and fixed `vermod.sh` and `reset_tstamp.sh`.  Tested this
code by changing the `MIN_TIMESTAMP` as noted above.  The `MIN_TIMESTAMP`
needed to up updated anyway due to changes in the `.info.json` and
`.author.json` formats.

Made numerous improvements and bug fixes to the Makefile.

Fixed how `picky` is used in by the `make picky` rule for a few special files.

Added multiple rules to the Makefile including but not limited to `make mkref`,
`make legacy_clobber`, and `make legacy_clean` rules.  Applied multiple bug
fixes to the Makefile.

Improved the Makefile to be less impacting for modern systems while trying to
maintain, for as long as we can, compatibility with some older systems.

Attempted to improve compatibility with reasonably modern systems.  We managed
to keep CentOS 7 somewhat supported for now, although we may be forced to drop
support of such an old system before 2024. But if we do reach 2024 we can only
say that given that CentOS 7 will reach EOL in 2024 and CentOS 8 support was
unduly dropped and since < CentOS 7 already was at EOL this should not be a
problem. :-)

Users of very out of date systems can still enter and submit entries to the
IOCCC.  They just might need to find a more modern system to package and submit
their IOCCC entry, however.

Added stub code for `hostchk.sh`.  A future release will include mode tests for
the given hosts.  Future releases will also include a bug report system that
will also use `hostchk.sh`.

Improved the `no-comment` directive in `.info.json` and `.author.json` files.

Improved how `time_t` values are used and printed.  We no longer assume that
`time_t` is signed nor assume it is unsigned: we only assume it's time. :-)

Improved comments in C code about special `seqcexit` comment directives.

Make numerous bug fixes and fixed a fair number of typos.  Nevertheless we claim

jnum_chk.c
/*
 * jnum_chk - tool to check JSON number string conversions
 *
 * "Because specs w/o version numbers are forced to commit to their original design flaws." :-)
 *
 * This JSON parser was co-developed by:
 *
 *	@xexyl
 *	https://xexyl.net		Cody Boone Ferguson
 *	https://ioccc.xexyl.net
 * and:
 *	chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
 *
 * "Because sometimes even the IOCCC Judges need some help." :-)
 *
 * "Share and Enjoy!"
 *     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
 */


/* special comments for the seqcexit tool */
/* exit code out of numerical order - ignore in sequencing - ooo */
/* exit code change of order - use new value in sequencing - coo */


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdint.h>
#include <inttypes.h>
#include <math.h>

/*
 * jnum_chk - tool to check JSON number string conversions
 */
#include "jnum_chk.h"

/*
 * version - official IOCCC toolkit versions
 */
#include "version.h"


/*
 * definitions
 */
#define REQUIRED_ARGS (0)	/* number of required arguments on the command line */


int
main(int argc, char *argv[])
{
    char const *program = NULL;	/* our name */
    extern char *optarg;	/* option argument */
    extern int optind;		/* argv index of the next arg */
    bool test = false;		/* true ==> JSON conversion test case result */
    bool error = false;		/* true ==> JSON conversion test suite error */
    bool strict = false;	/* true ==> strict testing for all struct integer element */
    struct json *node = NULL;	/* allocated JSON parser tree node */
    struct json_number *item = NULL;	/* JSON parser tree node for a JSON number */
    size_t len = 0;		/* length of str */
    int arg_count = 0;		/* number of args to process */
    int i;

    /*
     * parse args
     */
    program = argv[0];
    while ((i = getopt(argc, argv, "hv:J:VqS")) != -1) {
	switch (i) {
	case 'h':		/* -h - print help to stderr and exit 0 */
	    usage(3, program, "-h help mode", -1, -1); /*ooo*/
	    not_reached();
	    break;
	case 'v':		/* -v verbosity */
	    /*
	     * parse verbosity
	     */
	    verbosity_level = parse_verbosity(program, optarg);
	    break;
	case 'J':		/* -J json_verbosity */
	    /*
	     * parse JSON verbosity level
	     */
	    json_verbosity_level = parse_verbosity(program, optarg);
	    break;
	case 'V':		/* -V - print version and exit */
	    print("%s\n", JNUM_CHK_VERSION);
	    exit(3); /*ooo*/
	    not_reached();
	    break;
	case 'q':
	    quiet = true;
	    msg_warn_silent = true;
	    break;
	case 'S':		/* -S - strict mode */
	    strict = true;
	    break;
	default:
	    usage(4, program, "invalid -flag", -1, -1); /*ooo*/
	    not_reached();
	 }
    }
    arg_count = argc - optind;
    if (arg_count != REQUIRED_ARGS) {
	usage(4, program, "expected %d arguments, found: %d", REQUIRED_ARGS, arg_count); /*ooo*/
    }
    dbg(DBG_MED, "strict mode: %s", (strict == true) ? "enabled" : "disabled");

    /*
     * process each test case
     */
    for (i=0; i < test_count; ++i) {

	/*
	 * convert the test string
	 */
	node = json_conv_number_str(test_set[i], &len);
	if (node == NULL) {
	    warn(__func__, "json_conv_number_str(\"%s\", &%ju) returned NULL",
			   test_set[i], (uintmax_t)len);
	    error = true;
	    break;
	}
	if (node->type != JTYPE_NUMBER) {
	    err(19, program, "node->type for test %d: %d != %d", i, node->type, JTYPE_NUMBER);
	    not_reached();
	}
	item = &(node->item.number);

	/*
	 * compare conversion with test case
	 */
	test = chk_test(i, item, &test_result[i], len, strict);
	if (test == false) {
	    warn(__func__, "test: %d json_conv_number_str(\"%s\", &%ju) failed",
			   i, test_set[i], (uintmax_t)len);
	    error = true;
	}

	/*
	 * free converted node
	 */
	json_free(node, 0);
	if (node != NULL) {
	    free(node);
	    node = NULL;
	}
    }

    /*
     * All Done!!! - Jessica Noll, age 2
     */
    if (error == true) {
	if (strict == true) {
	    exit(2); /*ooo*/
	} else {
	    exit(1); /*ooo*/
	}
    }
    exit(0); /*ooo*/
}


/*
 * chk_test = check a given test
 *
 * When strict == true, we will compare every node->item.number element
 * with every *test element.  When strict == false, we will only compare
 * select struct json_number elements.
 *
 * given:
 *	testnum - test number being checked
 *	item	- pointer to struct json converted JSON number
 *	test	- pointer to test suite struct json_number
 *	len	- converted JSON number length
 *	strict	- true ==> compare all struct json_member elements,
 *		  false ==> compare only select struct json_member elements
 *
 * returns:
 *	true ==> check passed, no errors (under the given strict value) were found
 *	false ==> some check failed (under the given strict value), or
 *		  NULL pointer given, or len was incorrect
 */
static bool
chk_test(int testnum, struct json_number *item, struct json_number *test, size_t len, bool strict)
{
    bool test_error = false;	/* true ==> check failed */

    /*
     * firewall
     */
    if (item == NULL) {
	warn(__func__, "item is NULL");
	return false;
    }
    if (test == NULL) {
	warn(__func__, "test is NULL");
	return false;
    }
    if (len != test->number_len) {
	warn(__func__, "len: %ju != test->number_len: %ju", (uintmax_t)len, (uintmax_t)test->number_len);
	return false;
    }

    /*
     * test converted boolean
     */
    if (test_result[testnum].converted != item->converted) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].converted: %d != item.converted: %d",
		       testnum, test_result[testnum].converted, item->converted);
	test_error = true;
    } else {
	dbg(DBG_VVHIGH, "test_result[%d].converted: %d == item.converted: %d",
			testnum, test_result[testnum].converted, item->converted);
    }

    /*
     * test NULL strings
     */
    if (test_result[testnum].as_str == NULL) {
	if (item->as_str != NULL) {
	    dbg(DBG_VHIGH, "ERROR: test_result[%d].as_str == NULL && item->as_str != NULL", testnum);
	    test_error = true;
	} else {
	    dbg(DBG_VVHIGH, "OK: test_result[%d].as_str == NULL && item->as_str == NULL", testnum);
	}
    } else {
	/*
	 * test non-NULL strings
	 */
	if (item->as_str != NULL) {
	    dbg(DBG_VVHIGH, "OK: test_result[%d].as_str != NULL && item->as_str != NULL", testnum);
	} else {
	    dbg(DBG_VHIGH, "ERROR: test_result[%d].as_str != NULL && item->as_str == NULL", testnum);
	    test_error = true;
	}
    }

    /*
     * test string lengths
     */
    if (test_result[testnum].as_str_len != item->as_str_len) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].as_str_len: %ju != item->as_str_len: %ju",
		       testnum, (uintmax_t)test_result[testnum].as_str_len, (uintmax_t)item->as_str_len);
	test_error = true;
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].as_str_len: %ju == item->as_str_len: %ju",
		        testnum, (uintmax_t)test_result[testnum].as_str_len, (uintmax_t)item->as_str_len);
    }
    if (test_result[testnum].number_len != item->number_len) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].number_len: %ju != item->number_len: %ju",
		       testnum, (uintmax_t)test_result[testnum].number_len, (uintmax_t)item->number_len);
	test_error = true;
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].number_len: %ju == item->number_len: %ju",
		        testnum, (uintmax_t)test_result[testnum].number_len, (uintmax_t)item->number_len);
    }

    /*
     * test: other top level booleans
     */
    if (test_result[testnum].is_negative != item->is_negative) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].is_negative: %d != item.is_negative: %d",
		       testnum, test_result[testnum].is_negative, item->is_negative);
	test_error = true;
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].is_negative: %d == item.is_negative: %d",
			testnum, test_result[testnum].is_negative, item->is_negative);
    }
    if (test_result[testnum].is_floating != item->is_floating) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].is_floating: %d != item.is_floating: %d",
		       testnum, test_result[testnum].is_floating, item->is_floating);
	test_error = true;
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].is_floating: %d == item.is_floating: %d",
			testnum, test_result[testnum].is_floating, item->is_floating);
    }
    if (test_result[testnum].is_e_notation != item->is_e_notation) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].is_e_notation: %d != item.is_e_notation: %d",
		       testnum, test_result[testnum].is_e_notation, item->is_e_notation);
	test_error = true;
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].is_e_notation: %d == item.is_e_notation: %d",
			testnum, test_result[testnum].is_e_notation, item->is_e_notation);
    }

    /*
     * perform tests for both strict and non-strict testing of integers
     */

    /* test: int8_t */
    check_val(&test_error, "int8", testnum,
			   test_result[testnum].int8_sized, item->int8_sized,
			   test_result[testnum].as_int8, item->as_int8);

    /* test: uint8_t */
    check_uval(&test_error, "uint8", testnum,
			    test_result[testnum].uint8_sized, item->uint8_sized,
			    test_result[testnum].as_uint8, item->as_uint8);

    /* test: int16_t */
    check_val(&test_error, "int16", testnum,
			   test_result[testnum].int16_sized, item->int16_sized,
			   test_result[testnum].as_int16, item->as_int16);

    /* test: uint16_t */
    check_uval(&test_error, "uint16", testnum,
			    test_result[testnum].uint16_sized, item->uint16_sized,
			    test_result[testnum].as_uint16, item->as_uint16);

    /* test: int32_t */
    check_val(&test_error, "int32", testnum,
			   test_result[testnum].int32_sized, item->int32_sized,
			   test_result[testnum].as_int32, item->as_int32);

    /* test: uint32_t */
    check_uval(&test_error, "uint32", testnum,
			    test_result[testnum].uint32_sized, item->uint32_sized,
			    test_result[testnum].as_uint32, item->as_uint32);

    /* test: int64_t */
    check_val(&test_error, "int64", testnum,
			   test_result[testnum].int64_sized, item->int64_sized,
			   test_result[testnum].as_int64, item->as_int64);

    /* test: uint64_t */
    check_uval(&test_error, "uint64", testnum,
			    test_result[testnum].uint64_sized, item->uint64_sized,
			    test_result[testnum].as_uint64, item->as_uint64);

    /*
     * tests for strict mode only of integers
     */
    if (strict == true) {

	/* test: int */
	check_val(&test_error, "int", testnum,
			       test_result[testnum].int_sized, item->int_sized,
			       test_result[testnum].as_int, item->as_int);

	/* test: unsigned int */
	check_uval(&test_error, "uint", testnum,
				test_result[testnum].uint_sized, item->uint_sized,
				test_result[testnum].as_uint, item->as_uint);

	/* test: long */
	check_val(&test_error, "long", testnum,
			       test_result[testnum].long_sized, item->long_sized,
			       test_result[testnum].as_long, item->as_long);

	/* test: unsigned long */
	check_uval(&test_error, "ulong", testnum,
				test_result[testnum].ulong_sized, item->ulong_sized,
				test_result[testnum].as_ulong, item->as_ulong);

	/* test: long long */
	check_val(&test_error, "longlong", testnum,
			       test_result[testnum].longlong_sized, item->longlong_sized,
			       test_result[testnum].as_longlong, item->as_longlong);

	/* test: unsigned long long */
	check_uval(&test_error, "ulonglong", testnum,
				test_result[testnum].ulonglong_sized, item->ulonglong_sized,
				test_result[testnum].as_ulonglong, item->as_ulonglong);

	/* test: ssize_t */
	check_val(&test_error, "ssize", testnum,
			       test_result[testnum].ssize_sized, item->ssize_sized,
			       test_result[testnum].as_ssize, item->as_ssize);

	/* test: size_t */
	check_uval(&test_error, "size", testnum,
				test_result[testnum].size_sized, item->size_sized,
				test_result[testnum].as_size, item->as_size);

	/* test: off_t */
	check_val(&test_error, "off", testnum,
			       test_result[testnum].off_sized, item->off_sized,
			       test_result[testnum].as_off, item->as_off);
    }

    /*
     * more tests for both strict and non-strict testing of integers
     */

    /* test: intmax_t */
    check_val(&test_error, "intmax", testnum,
			   test_result[testnum].maxint_sized, item->maxint_sized,
			   test_result[testnum].as_maxint, item->as_maxint);

    /* test: uintmax_t */
    check_uval(&test_error, "uintmax", testnum,
			    test_result[testnum].umaxint_sized, item->umaxint_sized,
			    test_result[testnum].as_umaxint, item->as_umaxint);

    /*
     * perform tests for both strict and non-strict testing of floating point
     */

    /* test: float */
    check_fval(&test_error, "float", testnum,
			    test_result[testnum].float_sized, item->float_sized,
			    test_result[testnum].as_float, item->as_float,
			    test_result[testnum].as_float_int, item->as_float_int, strict);

    /* test: double */
    check_fval(&test_error, "double", testnum,
			    test_result[testnum].double_sized, item->double_sized,
			    test_result[testnum].as_double, item->as_double,
			    test_result[testnum].as_double_int, item->as_double_int, strict);

    /* test: long double */
    check_fval(&test_error, "longdouble", testnum,
			    test_result[testnum].longdouble_sized, item->longdouble_sized,
			    test_result[testnum].as_longdouble, item->as_longdouble,
			    test_result[testnum].as_longdouble_int, item->as_longdouble_int, strict);

    /*
     * if this test failed, force non-zero exit
     */
    if (test_error == true) {
	dbg(DBG_MED, "ERROR: test %d failed", testnum);
	return false;
    }
    dbg(DBG_HIGH, "OK: test %d passed", testnum);
    return true;
}


/*
 * check_val - test an aspect of conversion into struct json_number for signed integer
 *
 * If the reference matches the converted for a given test aspect, this
 * function just returns, else *testp is false.
 *
 * given:
 *	testp	- pointer to set to false if aspect of conversion fails
 *	type	- test aspect name type
 *	testnum	- test number
 *	size_a	- reference (test_result[i]) conversion into type boolean
 *	size_b	- converted (item->) type boolean
 *	val_a	- signed reference (test_result[i]) conversion into type value
 *	val_b	- signed reference (item->) conversion into type value
 *
 * NOTE: This function will warn in error.
 */
static void
check_val(bool *testp, char const *type, int testnum, bool size_a, bool size_b, intmax_t val_a, intmax_t val_b)
{
    /*
     * firewall
     */
    if (testp == NULL) {
	warn(__func__, "testp is NULL");
	return;
    }
    if (type == NULL) {
	warn(__func__, "type is NULL");
	return;
    }

    /*
     * compare booleans
     */
    if (size_a != size_b) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].%s_sized: %s != item->%s_sized: %s",
		       testnum,
		       type, booltostr(size_a),
		       type, booltostr(size_b));
	*testp = true; /* test failed */
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].%s_sized: %s == item->%s_sized: %s",
		        testnum,
		        type, booltostr(size_a),
		        type, booltostr(size_b));
    }

    /*
     * compare values
     */
    if (size_a == true && val_a != val_b) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].as_%s: %jd != item->as_%s: %jd",
		       testnum,
		       type, val_a,
		       type, val_b);
	*testp = true; /* test failed */
    } else if (size_a == true) {
	dbg(DBG_VVHIGH, "OK: test_result[%d].as_%s: %jd == item->as_%s: %jd",
		        testnum,
		        type, val_a,
		        type, val_b);
    }
    return;
}


/*
 * check_uval - test an aspect of conversion into struct json_number for unsigned integer
 *
 * If the reference matches the converted for a given test aspect, this
 * function just returns, else *testp is false.
 *
 * given:
 *	testp	- pointer to set to false if aspect of conversion fails
 *	type	- test aspect name type
 *	testnum	- test number
 *	size_a	- reference (test_result[i]) conversion into type boolean
 *	size_b	- converted (item->) type boolean
 *	val_a	- unsigned reference (test_result[i]) conversion into type value
 *	val_b	- unsigned reference (item->) conversion into type value
 *
 * NOTE: This function will warn in error.
 */
static void
check_uval(bool *testp, char const *type, int testnum, bool size_a, bool size_b, uintmax_t val_a, uintmax_t val_b)
{
    /*
     * firewall
     */
    if (testp == NULL) {
	warn(__func__, "testp is NULL");
	return;
    }
    if (type == NULL) {
	warn(__func__, "type is NULL");
	return;
    }

    /*
     * compare booleans
     */
    if (size_a != size_b) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].%s_sized: %s != item->%s_sized: %s",
		       testnum,
		       type, booltostr(size_a),
		       type, booltostr(size_b));
	*testp = true; /* test failed */
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].%s_sized: %s == item->%s_sized: %s",
		        testnum,
		        type, booltostr(size_a),
		        type, booltostr(size_b));
    }

    /*
     * compare values
     */
    if (size_a == true && val_a != val_b) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].as_%s: %ju != item->as_%s: %ju",
		       testnum,
		       type, val_a,
		       type, val_b);
	*testp = true; /* test failed */
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].as_%s: %ju == item->as_%s: %ju",
		        testnum,
		        type, val_a,
		       type, val_b);
    }
    return;
}


/*
 * check_fval - test an aspect of conversion into struct json_number for floating point value
 *
 * If the reference matches the converted for a given test aspect, this
 * function just returns, else *testp is false.
 *
 * given:
 *	testp	- pointer to set to false if aspect of conversion fails
 *	type	- test aspect name type
 *	testnum	- test number
 *	size_a	- reference (test_result[i]) conversion into type boolean
 *	size_b	- converted (item->) type boolean
 *	val_a	- signed reference (test_result[i]) conversion into type value
 *	val_b	- signed reference (item->) conversion into type value
 *	int_a	- reference (test_result[i]) is an integer
 *	int_b	- reference (item->) is an integer
 *	strict	- test for strict match
 *
 * NOTE: This function will warn in error.
 */
static void
check_fval(bool *testp, char const *type, int testnum, bool size_a, bool size_b,
	  long double val_a, long double val_b, bool int_a, bool int_b, bool strict)
{
    long double diff;		/* absolute difference between val_a and val_b */
    long double diff_part;	/* absolute difference between val_a and val_b as one part of val_a */

    /*
     * firewall
     */
    if (testp == NULL) {
	warn(__func__, "testp is NULL");
	return;
    }
    if (type == NULL) {
	warn(__func__, "type is NULL");
	return;
    }

    /*
     * compare booleans
     */
    if (size_a != size_b) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].%s_sized: %s != item->%s_sized: %s",
		       testnum,
		       type, booltostr(size_a),
		       type, booltostr(size_b));
	*testp = true; /* test failed */
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].%s_sized: %s == item->%s_sized: %s",
		        testnum,
		        type, booltostr(size_a),
		        type, booltostr(size_b));
    }
    if (int_a != int_b) {
	dbg(DBG_VHIGH, "ERROR: test_result[%d].%s_int: %s != item->%s_int: %s",
		       testnum,
		       type, int_a ? "true" : "false",
		       type, int_b ? "true" : "false");
	*testp = true; /* test failed */
    } else {
	dbg(DBG_VVHIGH, "OK: test_result[%d].%s_int: %s == item->%s_int: %s",
		        testnum,
		        type, int_a ? "true" : "false",
		        type, int_b ? "true" : "false");
    }

    /*
     * compare values only if size_a is true
     */
    if (size_a == true) {

	/*
	 * case strict: test for exact match
	 */
	if (strict == true) {
	    if (val_a != val_b) {
		dbg(DBG_VHIGH, "ERROR: test_result[%d].as_%s: %.22Lg != item->as_%s: %.22Lg",
			       testnum,
			       type, val_a,
			       type, val_b);
		*testp = true; /* test failed */
	    } else {
		dbg(DBG_VVHIGH, "OK: test_result[%d].as_%s: %.22Lg == item->as_%s: %.22Lg",
				testnum,
				type, val_a,
				type, val_b);
	    }

	/*
	 * case non-strict: test for match to 1 part in MATCH_PRECISION
	 */
	} else {

	    /* determine the difference */
	    diff = fabsl(val_a - val_b);

	    /* determine the difference as part of the whole */
	    if (val_a != 0.0) {
		diff_part = fabsl(diff / val_a);
	    } else {
		diff_part = diff;	/* zero value special case */
	    }

	    /* compare difference as part of the whole */
	    if (diff_part <= 1.0/MATCH_PRECISION) {
		dbg(DBG_VVHIGH, "OK: test_result[%d].as_%s: %.22Lg similar to item->as_%s: %.22Lg",
				testnum,
				type, val_a,
				type, val_b);
		dbg(DBG_VVHIGH, "OK: test_result[%d].as_%s: diff precision: %.22Lg <= 1.0/MATCH_PRECISION: %.22Lg",
			       testnum,
			       type, diff_part, 1.0/MATCH_PRECISION);

	    } else {
		dbg(DBG_VHIGH, "ERROR: test_result[%d].as_%s: %.22Lg not similar to item->as_%s: %.22Lg",
			       testnum,
			       type, val_a,
			       type, val_b);
		dbg(DBG_VHIGH, "ERROR: test_result[%d].as_%s: diff precision: %.22Lg > 1.0/MATCH_PRECISION: %.22Lg",
			       testnum,
			       type, diff_part, 1.0/MATCH_PRECISION);
		*testp = true; /* test failed */
	    }
	}
    }
    return;
}


/*
 * usage - print usage to stderr
 *
 * Example:
 *      usage(3, "missing required argument(s), program: %s", program);
 *
 * given:
 *	exitcode        value to exit with
 *	prog		our program name
 *	str		top level usage message
 *	expected	>= 0 ==> expected args, < 0 ==> ignored
 *	argc		>= 0 ==> argument count, < 0 ==> ignored
 *
 * NOTE: We warn with extra newlines to help internal fault messages stand out.
 *       Normally one should NOT include newlines in warn messages.
 *
 * This function does not return.
 */
static void
usage(int exitcode, char const *prog, char const *str, int expected, int argc)
{
    /*
     * firewall
     */
    if (prog == NULL) {
	prog = "((NULL prog))";
	warn(__func__, "\nin usage(): program was NULL, forcing it to be: %s\n", prog);
    }
    if (str == NULL) {
	str = "((NULL str))";
	warn(__func__, "\nin usage(): program was NULL, forcing it to be: %s\n", str);
    }

    /*
     * print the formatted usage stream
     */
    if (argc >= 0 && expected >= 0) {
	fprintf_usage(DO_NOT_EXIT, stderr, str, expected, argc);
    } else {
	fprintf_usage(DO_NOT_EXIT, stderr, "%s", str);
    }
    fprintf_usage(exitcode, stderr, usage_msg, prog, DBG_DEFAULT, JSON_DBG_DEFAULT, JNUM_CHK_VERSION);
    exit(exitcode); /*ooo*/
    not_reached();
}
